platforms/core-execution/workers/src/main/java/org/gradle/workers/internal/WorkerDaemonServer.java

```diff
@@ -179,6 +187,18 @@ protected ExecFactory createExecFactory(ExecFactory execFactory, FileResolver fi
                 .build();
         }
 
+        /**
+         * This is not correct! These services should be available in the build session scope, not the project scope.
+         * However, workers do not observe the same build lifecycle and do not stop or recreate build session services between builds.
+         * This works around that by recreating the build session scope services for every request.
+         */
+        BuildTreeScopedCacheBuilderFactory createBuildTreeScopedCache(UnscopedCacheBuilderFactory unscopedCacheBuilderFactory) {
+            return new DefaultBuildTreeScopedCacheBuilderFactory(projectCacheDir, unscopedCacheBuilderFactory);
+        }
+        DecompressionCache createDecompressionCacheFactory(BuildTreeScopedCacheBuilderFactory cacheBuilderFactory) {
+            return new DefaultDecompressionCache(cacheBuilderFactory);
+        }
+
```

Comment:
💭 Suggestion: `@Deprecated` these to make their incorrectness even more visible.

Reply:
Deprecated isn't quite the right message.

This is more like a warning that we're referring to these by a name (build tree scoped), but using them in a different way (per-request scope) because we can't honor their desired semantics (start once/stop once at end of build).

Reply:
I was thinking just to make it really obvious that something is wrong here, especially if you're reading the 2nd method and don't think the javadoc on the first applies to it.

Reply:
I reworked this isolate the special services into a separate provider that can be documented as unusual.